headeredevent.go

```diff
@@ -0,0 +1,84 @@
+package gomatrixserverlib
+
+import (
+	"encoding/json"
+	"reflect"
+
+	"github.com/tidwall/sjson"
+)
+
+// HeaderedEventHeader contains header fields for an event that contains
+// additional metadata, e.g. room version. IMPORTANT NOTE: All fields in
+// this struct must have a "json:" name tag or otherwise the reflection
+// code for marshalling and unmarshalling headered events will not work.
+type EventHeader struct {
+	RoomVersion RoomVersion `json:"room_version"`
+}
+
+// HeaderedEvent is a wrapper around an Event that contains information
+// about the room version. All header fields will be added into the event
+// when marshalling into JSON and will be separated out when unmarshalling.
+type HeaderedEvent struct {
+	EventHeader
+	Event
+}
+
+// UnmarshalJSON implements json.Unmarshaller
+func (e *HeaderedEvent) UnmarshalJSON(data []byte) error {
+	var err error
+	// First extract the headers from the JSON.
+	var m EventHeader
+	if err = json.Unmarshal(data, &m); err != nil {
```

Comment:
So doing this will needlessly cause `Unmarshal` to process all fields twice: once here and once on `:53`. I was hoping to see something like:
```
struct {
    Header EventHeader `json:"header"`
    Event json.RawMessage `json:"event"`
}
```
allowing the unmarshal to skip over the rest of the event, but we don't even namespace the header to allow for this operation. What's the rationale for sharing namespaces?

Reply:
If we do this then the struct format is different for marshalling vs unmarshalling, e.g. top-level `header` and `event` keys, which are not present currently. We could change the entire top-level `EventHeader` type but then all of the `Event` function calls throughout Dendrite also need to be updated to point to the `Event` field rather than sitting at the top level, so this meant that we could preserve as much of the current API shape as possible.

Reply:
> all of the Event function calls throughout Dendrite also need to be updated to point to the Event field rather than sitting at the top level

That seems reasonable, but it doesn't explain why we can't have a `Header` field to namespace the header/fields correctly e.g:
```
type HeaderedEvent struct {
    Header EventHeader
    Event
}
```

Reply:
So I've just given this a quick try and it adds more complexity into the unmarshaller, since it's then not possible to have a top-level embedded `Event` in the struct and still avoid unmarshalling it twice to separate the headers.

I think for now I'm leaning towards leaving this as it is, since in practice it's only really going to be used in a small number of cases. We could prefix the room version name tag or something, like `_room_version`, just to make it really obvious that it doesn't belong there if the headers do leak.

Reply:
Okay, JSON prefix with `_` SGTM for now.